tools/mypy_wrapper.py

```diff
@@ -18,47 +18,106 @@
 - https://github.com/pytorch/pytorch/wiki/Lint-as-you-type
 """
 
-import re
 import sys
+from collections import defaultdict
 from configparser import ConfigParser
-from pathlib import Path, PurePath
-from typing import List, Set
+from pathlib import Path, PurePath, PurePosixPath
+from typing import Any, Dict, List, Optional, Set, Tuple
 
 import mypy.api
 # not part of the public API, but this is the easiest way to ensure that
 # we agree with what mypy actually does
 import mypy.config_parser
 
 
-def config_files() -> Set[str]:
-    """
-    Return a set of the names of all the PyTorch mypy config files.
-    """
-    return {str(p) for p in Path().glob('mypy*.ini')}
+def read_config(config_path: Path) -> Set[str]:
+    config = ConfigParser()
+    config.read(config_path)
+    # hopefully on Windows this gives posix paths
+    return set(mypy.config_parser.split_and_match_files(
+        config['mypy']['files'],
+    ))
 
 
-def is_match(*, pattern: str, filename: str) -> bool:
-    """
-    Return True iff the filename matches the (mypy ini) glob pattern.
-    """
-    for path in mypy.config_parser.split_and_match_files(pattern):
-        path = PurePath(path).as_posix()
-        if filename == path or filename.startswith(f'{path}/'):
-            return True
-    return False
+def config_files() -> Dict[str, Set[str]]:
+    return {str(ini): read_config(ini) for ini in Path().glob('mypy*.ini')}
 
 
-def in_files(*, ini: str, py: str) -> bool:
-    """
-    Return True iff the py file is included in the ini file's "files".
-    """
-    config = ConfigParser()
+def split_path(path: str) -> List[str]:
+    pure = PurePosixPath(path)
+    return [str(p.name) for p in list(reversed(pure.parents))[1:] + [pure]]
+
+
+# mypy doesn't support recursive types yet
+# https://github.com/python/mypy/issues/731
+Trie = Dict[Optional[str], Any]
+
+
+def make_trie(configs: Dict[str, Set[str]]) -> Trie:
+    trie: Trie = {}
+    for ini, files in configs.items():
+        for f in files:
+            inner = trie
+            for segment in split_path(f):
+                inner = inner.setdefault(segment, {})
+            inner.setdefault(None, set()).add(ini)
+    return trie
+
+
+def lookup(trie: Trie, filename: str) -> Set[str]:
```

Comment:
Sure, I'll add comments to these functions describing their purpose 👍

> would it work if we turned all the non-specific files in the config into globs (i.e. append a `*`) and then use [fnmatch](https://docs.python.org/3/library/fnmatch.html)?

Maybe? It probably doesn't matter which approach we use since the lists of files are going to be very small, but I chose this trie approach because it has better asymptotic complexity than what you suggest (since this way you only have to do a lookup rather than iterate through everything every time). Again though, asymptotic complexity obviously doesn't matter here haha

Reply:
I generally err towards whatever is less code, especially if it's something from the standard library. These tools scripts especially aren't the most documented so people will probably come looking into the code to see what it does at some point, and the simpler it is the better imo